server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,636 @@
+        // TODO: FieldsReader has stats() which might contain the disk usage infos
+        // Also, can we track the byte reader per field extension to avoid visiting terms multiple times?
+        FieldsProducer postingsReader = reader.getPostingsReader();
+        if (postingsReader == null) {
+            return;
+        }
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetAndTrackBytesRead(BytesReadTrackingMode.SIZE_AND_POSITION);
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+            // Visit terms index and term dictionary
+            final long termBytes;
+            {
+                final BytesRefIterator termsIterator = terms.iterator();
+                final TermsEnum termsEnum = terms.iterator();
+                BytesRef bytesRef;
+                while ((bytesRef = termsIterator.next()) != null) {
+                    cancellationChecker.logEvent();
+                    final boolean found = termsEnum.seekExact(bytesRef);
+                    assert found;
+                }
+                termBytes = directory.getBytesRead();
+                if (termBytes > 0) {
+                    stats.addTerms(field.name, termBytes);
+                }
+            }
+
+            // Visit posting and skip lists
+            final long postingsBytes;
+            {
+                final TermsEnum termsEnum = terms.iterator();
+                while (termsEnum.next() != null) {
+                    termsEnum.totalTermFreq();
+                    postings = termsEnum.postings(postings, PostingsEnum.NONE);
+                    while (postings.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        cancellationChecker.logEvent();
+                        postings.freq();
+                    }
+                    // skip lists
+                    for (long idx = 1; idx <= 8; idx++) {
+                        cancellationChecker.logEvent();
+                        final int skipDocID = Math.toIntExact(idx * reader.maxDoc() / 8);
+                        postings = termsEnum.postings(postings, PostingsEnum.NONE);
+                        if (postings.advance(skipDocID) != DocIdSetIterator.NO_MORE_DOCS) {
+                            postings.freq();
+                            postings.nextDoc();
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                postingsBytes = directory.getBytesRead() - termBytes;
+                if (postingsBytes > 0) {
+                    stats.addPosting(field.name, postingsBytes);
```

Comment:
why distinguish terms and postings? Can you have one without the other?

Reply:
I'm happy either way, merging both might help avoid some confusion with keyword fields, which inline the matching docID into the terms dict when docFreq=1. So in such cases, you could have megabytes (or even gigabytes for unique fields) attributed to terms that would actually be due to postings.

Reply:
I'm currently not distinguishing between these in the fields usage API (not sure yet how), therefore have a slight preference for combining them.

Reply:
Ok, I will combine these fields :)